PyInstaller/building/build_main.py

```diff
@@ -389,71 +407,73 @@ def assemble(self):
-        #       to module which is reachable for top-level script. Or if these
-        #       files belong to a module from dead branch of the graph.
+        # 1. All hook() functions defined in cached hooks for imported modules
+        #    are called. This may result in new modules being imported (e.g., as
+        #    hidden imports) that were ignored earlier in the current iteration:
+        #    if this is the case, all hook() functions defined in cached hooks
+        #    for these modules will be called by the next iteration.
+        # 2. All cached hooks whose hook() functions were called are removed
+        #    from this cache. If this cache is empty, no hook() functions will
+        #    be called by the next iteration and this loop will be terminated.
+        # 3. If no hook() functions were called, this loop is terminated.
+        logger.info('Loading module hooks...')
+
+        # Cache of all external dependencies (e.g., binaries, datas) listed in
+        # hook scripts for imported modules.
         additional_files_cache = AdditionalFilesCache()
 
+        # For each imported module, run this module's post-graph hooks if any.
         while True:
-            # This ensures that import hooks get applied only once.
-            applied_hooks = []  # Empty means no hook was applied.
-
-            # Iterate over hooks in cache.
-            for imported_name in hooks_cache:
-
-                # Skip hook if no module for it is in the graph.
-                from_node = self.graph.findNode(imported_name, create_nspkg=False)
-                if from_node is None:
+            # Set of the names of all imported modules whose post-graph hooks
+            # are run by this iteration, preventing the next iteration from re-
+            # running these hooks. If still empty at the end of this iteration,
+            # no post-graph hooks were run; thus, this loop will be terminated.
+            hooked_module_names = set()
+
+            # For each remaining hookable module and corresponding hooks...
+            for module_name, module_hooks in module_hook_cache.items():
+                # Graph node for this module if imported or "None" otherwise.
+                module_node = self.graph.findNode(
+                    module_name, create_nspkg=False)
+
+                # If this module has not been imported, temporarily ignore it.
+                # This module is retained in the cache, as a subsequently run
+                # post-graph hook could import this module as a hidden import.
+                if module_node is None:
                     continue
-                # Skip hook if not the right Node type.
-                node_type = type(from_node).__name__
-                if node_type not in VALID_MODULE_TYPES:
+
+                #FIXME: This is... wierd. What's this about again?
+                # If this module is not a module at all, permanently ignore it.
```

Comment:
I believe this is because `findNode()` is allowed to return things like `MissingModule` or `InvalidCompiledModule`, which we are not interested in running hooks for. Mostly, `MissingModule` can be created for a name that is imported from a module, as in `from foo import bar`, so we don't want those.